src/dbnode/persist/fs/types.go

```diff
@@ -325,6 +347,63 @@ type IndexFileSetReader interface {
 	Validate() error
 }
 
+// FileStatus describes the current file status.
+type FileStatus struct {
+	Type FileStatusType
+}
+
+// FileStatusType is a type of file status
+type FileStatusType uint
+
+const (
+	// AvailableFileStatusType describes an available file (exists/readable).
+	AvailableFileStatusType FileStatusType = iota
+	// RemovingFileStatusType describes a file that is being removed and
+	// waiting for ref counts to close.
+	RemovingFileStatusType
+	// RemovedFileStatusType describes a file that is deleted after
+	// ref counts have closed.
+	RemovedFileStatusType
+)
+
+// File is a the os.File methods that wraps the close and provides
+// ref counting for file deletion.
+type File interface {
+	// Status methods.
+	Status() FileStatus
+	PathAbs() string
+
+	// Watchers.
+	WatchStatus(watcher FileStatusWatcher) close.SimpleCloser
+
+	// os.File methods.
+	Close() error
+	Fd() uintptr
+	Name() string
+	Read(b []byte) (int, error)
+	ReadAt(b []byte, off int64) (int, error)
+	Readdir(n int) ([]os.FileInfo, error)
+	Readdirnames(n int) ([]string, error)
+	Seek(offset int64, whence int) (ret int64, err error)
+	Stat() (os.FileInfo, error)
+	Sync() error
+	Write(b []byte) (int, error)
+	WriteAt(b []byte, off int64) (int, error)
+}
+
+// FileStatusWatcher implements a watcher of file status.
+type FileStatusWatcher interface {
+	OnFileStatusChange(f File, status FileStatus)
+}
+
+// FileRegistry is a file lifecycle manager and should always be used in place
+// of os.Open to ensure
+type FileRegistry interface {
```

Comment:
Is `Filesystem` a better name/abstraction? 

Would be better to avoid mixing other FD calls between this type and the os package. e.g. chmod/stat/rename/etc. Can be as simple as pass thru for now. 

Down the road, this would let us do nice things like mock the filesystem used during tests. All manner of benefits including speed/injection/etc.

Reply:
Found another package abstracting a FS here: https://github.com/spf13/afero#list-of-all-available-functions

Reply:
I think mocking the entire filesystem would be a lot more involved than just what this P.R is doing no? We still need to be able to list all directories in a folder, etc etc